src/Microsoft.AspNetCore.Routing/Matchers/CandidateSet.cs

```diff
@@ -2,56 +2,149 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.AspNetCore.Routing.Matchers
 {
-    internal class CandidateSet
+    public sealed class CandidateSet
     {
-        public static readonly CandidateSet Empty = new CandidateSet(Array.Empty<Candidate>(), Array.Empty<int>());
-
-        // The array of candidates.
-        public readonly Candidate[] Candidates;
-
-        // The number of groups.
-        public readonly int GroupCount;
-
-        // The array of groups. Groups define a contiguous sets of indices into
-        // the candidates array.
-        //
-        // The groups array always contains N+1 entries where N is the number of groups.
-        // The extra array entry is there to make indexing easier, so we can lookup the 'end'
-        // of the last group without branching.
-        //
-        // Example:
-        //    Group0: Candidates[0], Candidates[1]
-        //    Group1: Candidates[2], Candidates[3], Candidates[4]
-        //
-        // The groups array would look like: { 0, 2, 5, }
-        public readonly int[] Groups;
-
-        public CandidateSet(Candidate[] candidates, int[] groups)
+        private CandidateState _state0;
+        private CandidateState _state1;
+        private CandidateState _state2;
+        private CandidateState _state3;
```

Comment:
4 is a number I picked.

Reply:
I'm going to document that it's a guess for now. We can collect data on this but I don't think it's entirely pressing.

Reply:
Generally how this will come into play is for a controller that has 4 actions with different verbs and the same route template.